### ISSUE TITLE
Not correctly understanding the Multi Head Attention part of the ViT implementation...

### ISSUE THREAD

**Issue by JavierUrenaPhDProjects** [lucidrains/vit-pytorch#282]
So I've been checking the ViT and Simple ViT implementations and so far they are very readable and easy to understand. But I dont quite comprehend what is being done here, in comparison with what the original ViT paper explains. Lets check the ViT implementation. This is part of the __init__ of the model class:

`inner_dim = dim_head *  heads`
`...`
`self.to_qkv = nn.Linear(dim, inner_dim * 3, bias = False)`
`self.to_out = nn.Sequential(
            nn.Linear(inner_dim, dim),
            nn.Dropout(dropout)
        ) if project_out else nn.Identity()`

As far as I understand here, explaining the forward from here: inner dim is like the "inner dimension" of the MSA block. For example, if each att head has a dim of 64, and there are 16 heads, then the inner dimension is 64*16 would be 1024. Then, with this what it does in `to_qkv` is projecting linearly the patch embedding (of dimension D, which is 1024 in the paper) from D to inner_dim * 3, so later that new extended embedding can be divided into 3 (query, key and value, all of them of equal dimension = inner_dim / 3). After doing all the necessary operations with the k, q, v, it does a new linear projection, transforming from inner_dim to D back.

My question is... is this really MSA? are we just linearly projecting to a greater dimension specified by hyperparameters and then projecting back to the embedding dimension? Maybe I'm wrong and this is a simplified way of achieving MSA, but, seems kinda weird. torch.nn has actually the MultiHeadAttention. I haven't checked that implementation thoroughly but it seems much vaster than just this one. Can somebody confirm me why this method is equivalent to a MSA? And sorry for ignorance

**Comment by stas-sl** [comment]
> is this really MSA? are we just linearly projecting to a greater dimension specified by hyperparameters and then projecting back to the embedding dimension?

For some reason you missed those lines from `Attention.forward`:
```python
qkv = self.to_qkv(x).chunk(3, dim = -1)
q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> b h n d', h = self.heads), qkv)

dots = torch.matmul(q, k.transpose(-1, -2)) * self.scale

attn = self.attend(dots)
attn = self.dropout(attn)

out = torch.matmul(attn, v)
```

So it is not just linear projection to a greater dimension and back. Basically it does exactly what attention should do according to formula: 
![image](https://github.com/lucidrains/vit-pytorch/assets/4602302/e2b1e69b-5666-4b76-a939-14e711f789d1)

First, it computes similarity between keys and queries, then takes softmax (non linearity) to get attention weights and then applies them to values. 

pytorch's `MultiHeadAttention` is definitely more verbose, it supports various mask, which are needed if you need causality  (decoder) or you have sequences of different length, or it could use some cuda optimized hardware implementation, but in the simplest form that's basically the attention.

**Comment by lucidrains** [comment]
@stas-sl thank you Stas

@JavierUrenaPhDProjects i would not recommend pytorch's MHA. you will quickly find it inflexible if you want to do anything out of the ordinary

**Comment by lucidrains** [comment]
@JavierUrenaPhDProjects the basic multi-head attention can be done in less than 10 loc with einops. i recommend you just recite it until you know it by heart, due to its current importance. masking is really what is complex when engineering with attention, but luckily you don't need to do much masking when working with ViTs.

**Comment by lucidrains** [comment]
closing because not an issue
